### ISSUE TITLE
Add post-removal notification for Optimize

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#6470]
## Feature Description

For sites that seem to have the Optimize module connected after it is removed from Site Kit, we will show a final notification about the removal to ensure admins are aware.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

**Note: This change is to be made after #6469 but included in the same release**

* If the Optimize module is active, and the module appears to be in a connected state (i.e. its settings have a non-empty `optimizeID`), a banner notification should be displayed on the main authenticated Site Kit dashboard. 
Note: these will have to be checked directly as the module's infrastructure will no longer exist.

**Title**
> Google Optimize has been removed

**Body**
> Google Optimize is no longer available as of September 30, 2023 and has since been removed from Site Kit.

**Primary CTA** ("OK, Got it!")
When clicking the primary CTA to dismiss the banner, the Optimize module should be "deactivated" and its module settings removed, effectively dismissing the banner permanently.

**Learn More URL**
```
https://support.google.com/optimize/answer/12979939
```

* In terms of priority, the banner should be displayed just above zero data state notifications, but after anything else above that  

#### Create a class to handle Optimize removal logic

Create a new class `Optimize_Sunset` in `includes/Modules/Optimize` with the following:
* Implement a static method `is_optimize_connected()` to check if Optimize is connected and active.
    * This method should use the `Options::get()` method by passing `googlesitekit_optimize_settings` to retrieve the Optimize settings and determine connectivity.
* Use the `googlesitekit_inline_base_data` filter hook to expose the Optimize connectivity data on the page load using the above method.
* Add a new POST REST API endpoint `deactivate-optimize` using the `REST_Route` constructor.
* It should remove the Optimize module settings using the `Options::delete()` method by passing `googlesitekit_optimize_settings`.
* The endpoint should return `success` with a value of `true` if the module was deactivated successfully.

#### Create client-side fetch stores

In `assets/js/googlesitekit/datastore/site/info.js`:
* Add a new selector, `checkOptimizeConnected`, which returns the `optimizeConnected` value from the inline base data.

In `assets/js/googlesitekit/modules/datastore/modules.js`:
* Add a new `fetchDeactivateOptimize` fetch store using `createFetchStore` with the following:
    * `baseName` should be `deactivateOptimize`.
    * `controlCallback` returns the `API.set( 'core', 'modules', 'deactivate-optimize' )` call.
    * `reducerCallback` receives `state` and `response` and returns an object with the existing `state` and a new `optimizeConnected` property set to the `false` value.
* Add a new action, `*deactivateOptimize`, which calls `fetchDeactivateOptimize` that dispatches the `receiveDeactivateOptimize` action from the fetch store.

#### Create a notification component

* Create a new component, `OptimizeModuleRemovedNotification` in `assets/js/components/notifications` with the following:
  * Check whether the Optimize module is connected using the new `checkOptimizeConnected` selector.
  * Return `null` if the selector returns a falsy value.
  * Otherwise, render the `BannerNotification` component with the following props:
    * `id`: `optimize-module-removed-notification`.
    * `title` and `description`: Translatable strings according to the AC.
    * `ctaLabel`: Translatable string "OK, Got it!".
    * `onCTAClick`: It should dispatch the `deactivateOptimize` action within a callback function.
    * `learnMoreLabel`: Translatable string "Learn more".
    * `learnMoreURL`: Translatable string `https://support.google.com/optimize/answer/12979939`.
* Render the `OptimizeModuleRemovedNotification` component within `BannerNotifications` just above the `ZeroDataStateNotifications` banner. 
* Add storybook stories for the `OptimizeModuleRemovedNotification` notification banner.

### Test Coverage

* Add PHP tests for the `Optimize_Sunset` class methods.
* Add unit tests for the `OptimizeModuleRemovedNotification` component.

## QA Brief

* 

## Changelog entry

*

**Comment by felixarntz** [comment]
@aaemnnosttv Maybe worth clarifying that the Optimize module won't exist anymore at this point? It sounds potentially misleading not having that here. The options will need to be more or less "manually" need to be retrieved from the database, outside of relying on anything related to the actual `Optimize` module implementation. I think that's worth clarifying in the ACs.

**Comment by aaemnnosttv** [comment]
Thanks @felixarntz !

> Maybe worth clarifying that the Optimize module won't exist anymore at this point? It sounds potentially misleading not having that here.

Hmmm, I'm open to any suggestions about the wording if you have any. Perhaps for this notification specifically we should link to a new SK support doc, rather than the one about the sunset of the service itself. I think it justifies its own doc.

> The options will need to be more or less "manually" need to be retrieved from the database, outside of relying on anything related to the actual `Optimize` module implementation. I think that's worth clarifying in the ACs.

It is mentioned where it says `Note: these will have to be checked directly as the module's infrastructure will no longer exist.` Do you feel there is anything more to clarify there or maybe you missed it?

**Comment by felixarntz** [comment]
@aaemnnosttv Ah I missed that indeed. My entire comment was focused on that point, so all good then! ✅

**Comment by marrrmarrr** [comment]
The current wording sgtm. We can reconsider adding a link to alternatives, if GA comes up with something new by then (as they seemed to indicate in the email about the end of Optimize).

**Comment by aaemnnosttv** [comment]
Thanks @marrrmarrr! Moving this over into IB 👍

**Comment by tofumatt** [comment]
We already have a module deactivation REST API endpoint/[deactivate_module function](https://github.com/google/site-kit-wp/blob/ab7d43767d14a5d9282db3acef91b105778d7a8f/includes/Core/Modules/Modules.php#L693), so I think creating a whole new endpoint, fetch store, etc. is a lot of duplication and extra effort.

We have a `Module_With_Deactivation` class that implements a `on_deactivation` feature I think we could use to handle the extra "cleanup"/"sunsetting" of Optimize that we want here without needing to create a whole new endpoint.

The module will still exist internally, right? Otherwise we can't use `deactivate_module` either since it won't exist, I think?

If it's possible to just use the existing deactivation infrastructure I think we should do that. Is there a reason we can't?

**Comment by hussain-t** [comment]
@tofumatt, Optimize module's infrastructure will be removed in #6469. The client-side `deactivateModule` action calls the `activation` endpoint. And the REST controller checks if the module is registered/available in [`Modules::get_module()`](https://github.com/google/site-kit-wp/blob/fa1d07f81ec790176f84fba3e88275f73a1e64b7/includes/Core/Modules/Modules.php#L531-L540). Because Optimize won't be available in `Modules:get_available_modules()`. Hence the [API](https://github.com/google/site-kit-wp/blob/fa1d07f81ec790176f84fba3e88275f73a1e64b7/includes/Core/Modules/REST_Modules_Controller.php#L250-L254) will throw an `invalid_module_slug` 500 (Internal Server Error) error. Please correct me if I need to include anything. Thanks!

cc: @aaemnnosttv

**Comment by tofumatt** [comment]
@hussain-t Makes sense, but will `Modules::deactivate_module( 'optimize' )` still work in that case? If the infrastructure is all removed, won't this line fail: https://github.com/google/site-kit-wp/blob/ab7d43767d14a5d9282db3acef91b105778d7a8f/includes/Core/Modules/Modules.php#L695?

**Comment by hussain-t** [comment]
You're right. Thanks, @tofumatt. I have removed that part from the IB.

**Comment by tofumatt** [comment]
The ACs here mention checking not only for non-empty `optimizeID`, but also for the module being "activated" and "deactivating" it if the "OK, Got it!" button is clicked.

I think the IB here covers everything and should work fine, but _technically_ they don't "deactivate" the module—which I assume means calling [`$options->set( self::OPTION_ACTIVE_MODULES, $active_modules_without_optimize );`](https://github.com/google/site-kit-wp/blob/3be399119f5941d93976c289349040c217e0aeaf/includes/Core/Storage/Options.php#L92).

@aaemnnosttv Is that required/part of the ACs? The wording of "active" and "deactivate" here is just a bit unclear so I wanted to check. I feel like it'd be good to remove an unused module from that list, but then I feel like that would be better to do elsewhere (as part of a migration when it's actually removed in #6469), so it doesn't need to be here.

If this is really just about the Optimize options being non-empty, we should adjust the ACs just so it's very clear that's all we care about here. Could you clarify @aaemnnosttv?

**Comment by hussain-t** [comment]
I agree with @tofumatt. The below section from the AC is a bit confusing:
>When clicking the primary CTA to dismiss the banner, the Optimize module should be "deactivated" and its module settings removed, effectively dismissing the banner permanently.

Because the infrastructure is removed in #6469, technically, we cannot deactivate a module that doesn't exist.

**Comment by aaemnnosttv** [comment]
> I think the IB here covers everything and should work fine, but _technically_ they don't "deactivate" the module—which I assume means calling [`$options->set( self::OPTION_ACTIVE_MODULES, $active_modules_without_optimize );`](https://github.com/google/site-kit-wp/blob/3be399119f5941d93976c289349040c217e0aeaf/includes/Core/Storage/Options.php#L92).
> Is that required/part of the ACs?

@tofumatt Yes, this is important to prevent the notification from showing again as mentioned in the primary CTA:  
- > When clicking the primary CTA to dismiss the banner, the Optimize module should be "deactivated" and its module settings removed, effectively dismissing the banner permanently.

> Create a REST API endpoint to check Optimize is connected

This isn't really necessary here, we can simply provide this value on page load. We should minimize the amount of temporary infra needed here.

> `OptimizeConnectedNotification`

This notification is only shown when Optimize seems to have been in a connected state, but the notification itself is about Optimize being removed so we should name it accordingly. Something like `OptimizeModuleRemovedNotification` would be more appropriate I think.

---

One other thing to keep in mind here is simplifying the eventual removal of these temporary parts. On the backend, perhaps a single `Optimize_Sunset` class would be good to encapsulate everything we need here. For data exposed to the frontend, it can hook into existing filters for inline data. It can work with an `Options` instance directly. Later, this should be something we can remove by simply deleting a few files and removing line or two where they're integrated.

Similarly, this should be a relatively simple implementation. The current (largest) estimate we have shouldn't be necessary and indicates we need to see where the approach can be simplified.

**Comment by aaemnnosttv** [comment]
> * Use the `googlesitekit_inline_base_data` filter hook to expose the Optimize connectivity data on the page load using the above method.

The base data isn't the best place for this since this is more for minimum data that SK would need on any screen/context.

---

After giving this some more thought, I think reducing Optimize to the minimal footprint requires keeping the module to some extent. Essentially, the module could be made `internal` so that it would no longer appear in any interfaces, but would still exist so that we could use the core APIs for checking whether it's connected, and deactivating it rather than adding new infrastructure to do this in a roundabout way.

I realize this would somewhat contradict the AC for #6469 but we could also make this issue about reintroducing the "internal/placeholder" version of the module instead, which would be almost nothing.

Alternatively, we could implement this notification on the service which wouldn't require that we keep anything of Optimize at all. Let me look into this as a first step and then we can finalize the direction here. If we decide to go the service-route, then this issue shouldn't be needed at all.

**Comment by bethanylang** [comment]
Per @eugene-manuilov: 

> yes, we can implement it on the service side. We have historical data in the BigQuery table. Shouldn't be very hard as far as I see it.

@aaemnnosttv Let me know if we're okay to close this out and if so I can add a task on the service side for this and loop you and Eugene in.

**Comment by aaemnnosttv** [comment]
Thanks @bethanylang – let's plan on doing this on the service side to avoid needing another issue in the future to remove the things we added about removing Optimize 😄
